vim_onenote.ahk

```diff
@@ -45,65 +49,160 @@ IsLastKey(key)
     return (A_PriorHotkey == key and A_TimeSincePriorHotkey < 400)
 }
 
-; ESC enters Normal Mode
-ESC::
-	Suspend, Off
-    ; send escape on through to exit from find
-    SendInput, {ESC} 
-	ToolTip, OneNote Vim Command Mode Active, 0, 0
-return
+
 
 ;--------------------------------------------------------------------------------
 ; Return to InsertMode
 InsertMode:
     ToolTip
     Suspend, On
+    hotkey, ESC, on
 return
 
 ;--------------------------------------------------------------------------------
+
+;  ctrl + [, ESC enter Normal Mode.
+; Not using ctrl + c, as people may still want to use for copying.
+;^c::
+^[::
+ESC::
+    ; Is not affected by insertmode's suspend
+    suspend, permit
+    gosub NormalMode
+return
+
+; imap workings (eg jj) currently can't be implemented because of how 
+; insert mode works.
+
+
+NormalMode:
+    Suspend, Off
+    ToolTip, OneNote Vim Command Mode Active, 0, 0
+return
+
+
+;--------------------------------------------------------------------------------
 +i::
-    SendInput, {Home}
+    send {Home}
     Gosub InsertMode
 return 
 
 i::Gosub InsertMode
 return 
 
 ;--------------------------------------------------------------------------------
+PrepareClipboard(){
+    ; push clipboard to local variable
+    ClipSaved := ClipboardAll
+    ; Clear clipboard to avoid errors
+    Clipboard :=
+}
+
+Copy(){
+    PrepareClipboard()
+    send ^c
+    ClipWait
+}
+
+Paste(){
+    Send %Clipboard%
+    RestoreClipboard()
+}
+
+RestoreClipboard(){
+    ;restore original clipboard
+    Clipboard := ClipSaved
+    ;ClipWait
+    ClipSaved := ; free memory
+}
+
+GetCursorColumn(){
+    send +{home}
+    Copy()
+    position := strLen(Clipboard)
+    RestoreClipboard()
+    return position
+}
+
+ConvertMotionToFunctionName(letter){
+    ;StringLower, test, letter
+    ;msgbox, %test% 
+    if letter is upper
+        return s%letter%
+    else if letter = 0
+        return z
+    else
+        return letter
+}
+
+SelectMotion(){
+    gosub, InsertMode
+    Input, motion, L1
+    gosub, NormalMode
+    ;if motion = i, a or digit, need to wait. If digit, loop motion that many times. If i, g or w, wait for anotther motion.
+    if IsLastKey(motion){
+        send {end}
+        send +{home}
+        return
+    }
+    MoveFunction := ConvertMotionToFunctionName(motion)
+    ;MsgBox, %MoveFunction%
+    send {shift down}
+    %MoveFunction%()
+    send {shift up}
+}
+
+
+w(){
+    send ^{right}
+}
+w::w()
```

Comment:
Part of the generic motion logic. Making the actual motion a function allows the motion to be called after a d or c, etc, within InputMotionAndMakeItSelection(). Then the hotkey calls the function, rather than directly making the motion.